// angular2-notifications/src/simple-notifications/components/notification.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@angular/core");
const animations_1 = require("@angular/animations");
const platform_browser_1 = require("@angular/platform-browser");
const notifications_service_1 = require("../services/notifications.service");
let NotificationComponent = class NotificationComponent {
    constructor(notificationService, domSanitizer, zone) {
        this.notificationService = notificationService;
        this.domSanitizer = domSanitizer;
        this.zone = zone;
        // Progress bar variables
        this.progressWidth = 0;
        this.stopTime = false;
        this.count = 0;
        this.instance = () => {
            this.zone.runOutsideAngular(() => {
                this.zone.run(() => this.diff = (new Date().getTime() - this.start) - (this.count * this.speed));
                if (this.count++ === this.steps)
                    this.zone.run(() => this.remove());
                else if (!this.stopTime) {
                    if (this.showProgressBar)
                        this.zone.run(() => this.progressWidth += 100 / this.steps);
                    this.timer = setTimeout(this.instance, (this.speed - this.diff));
                }
            });
        };
    }
    ngOnInit() {
        if (this.animate) {
            this.item.state = this.animate;
        }
        if (this.item.override) {
            this.attachOverrides();
        }
        if (this.timeOut !== 0) {
            this.startTimeOut();
        }
        this.safeSvg = this.domSanitizer.bypassSecurityTrustHtml(this.icon || this.item.icon);
    }
    startTimeOut() {
        this.steps = this.timeOut / 10;
        this.speed = this.timeOut / this.steps;
        this.start = new Date().getTime();
        this.zone.runOutsideAngular(() => this.timer = setTimeout(this.instance, this.speed));
    }
    onEnter() {
        if (this.pauseOnHover) {
            this.stopTime = true;
        }
    }
    onLeave() {
        if (this.pauseOnHover) {
            this.stopTime = false;
            setTimeout(this.instance, (this.speed - this.diff));
        }
    }
    setPosition() {
        return this.position !== 0 ? this.position * 90 : 0;
    }
    onClick($e) {
        this.item.click.emit($e);
        if (this.clickToClose) {
            this.remove();
        }
    }
    // Attach all the overrides
    attachOverrides() {
        Object.keys(this.item.override).forEach(a => {
            if (this.hasOwnProperty(a)) {
                this[a] = this.item.override[a];
            }
        });
    }
    ngOnDestroy() {
        clearTimeout(this.timer);
    }
    remove() {
        if (this.animate) {
            this.item.state = this.animate + 'Out';
            this.zone.runOutsideAngular(() => {
                setTimeout(() => {
                    this.zone.run(() => this.notificationService.set(this.item, false));
                }, 310);
            });
        }
        else {
            this.notificationService.set(this.item, false);
        }
    }
};
__decorate([
    core_1.Input(),
    __metadata("design:type", Number)
], NotificationComponent.prototype, "timeOut", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", Boolean)
], NotificationComponent.prototype, "showProgressBar", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", Boolean)
], NotificationComponent.prototype, "pauseOnHover", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", Boolean)
], NotificationComponent.prototype, "clickToClose", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", Number)
], NotificationComponent.prototype, "maxLength", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", String)
], NotificationComponent.prototype, "theClass", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", Boolean)
], NotificationComponent.prototype, "rtl", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", String)
], NotificationComponent.prototype, "animate", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", Number)
], NotificationComponent.prototype, "position", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", Object)
], NotificationComponent.prototype, "item", void 0);
NotificationComponent = __decorate([
    core_1.Component({
        selector: 'simple-notification',
        encapsulation: core_1.ViewEncapsulation.None,
        animations: [
            animations_1.trigger('enterLeave', [
                // Enter from right
                animations_1.state('fromRight', animations_1.style({ opacity: 1, transform: 'translateX(0)' })),
                animations_1.transition('* => fromRight', [
                    animations_1.style({ opacity: 0, transform: 'translateX(5%)' }),
                    animations_1.animate('400ms ease-in-out')
                ]),
                animations_1.state('fromRightOut', animations_1.style({ opacity: 0, transform: 'translateX(-5%)' })),
                animations_1.transition('fromRight => fromRightOut', [
                    animations_1.style({ opacity: 1, transform: 'translateX(0)' }),
                    animations_1.animate('300ms ease-in-out')
                ]),
                // Enter from left
                animations_1.state('fromLeft', animations_1.style({ opacity: 1, transform: 'translateX(0)' })),
                animations_1.transition('* => fromLeft', [
                    animations_1.style({ opacity: 0, transform: 'translateX(-5%)' }),
                    animations_1.animate('400ms ease-in-out')
                ]),
                animations_1.state('fromLeftOut', animations_1.style({ opacity: 0, transform: 'translateX(5%)' })),
                animations_1.transition('fromLeft => fromLeftOut', [
                    animations_1.style({ opacity: 1, transform: 'translateX(0)' }),
                    animations_1.animate('300ms ease-in-out')
                ]),
                // Rotate
                animations_1.state('scale', animations_1.style({ opacity: 1, transform: 'scale(1)' })),
                animations_1.transition('* => scale', [
                    animations_1.style({ opacity: 0, transform: 'scale(0)' }),
                    animations_1.animate('400ms ease-in-out')
                ]),
                animations_1.state('scaleOut', animations_1.style({ opacity: 0, transform: 'scale(0)' })),
                animations_1.transition('scale => scaleOut', [
                    animations_1.style({ opacity: 1, transform: 'scale(1)' }),
                    animations_1.animate('400ms ease-in-out')
                ]),
                // Scale
                animations_1.state('rotate', animations_1.style({ opacity: 1, transform: 'rotate(0deg)' })),
                animations_1.transition('* => rotate', [
                    animations_1.style({ opacity: 0, transform: 'rotate(5deg)' }),
                    animations_1.animate('400ms ease-in-out')
                ]),
                animations_1.state('rotateOut', animations_1.style({ opacity: 0, transform: 'rotate(-5deg)' })),
                animations_1.transition('rotate => rotateOut', [
                    animations_1.style({ opacity: 1, transform: 'rotate(0deg)' }),
                    animations_1.animate('400ms ease-in-out')
                ])
            ])
        ],
        template: `
        <div class="simple-notification"
            [@enterLeave]="item.state"
            (click)="onClick($event)"
            [class]="theClass"

            [ngClass]="{
                'alert': item.type === 'alert',
                'error': item.type === 'error',
                'success': item.type === 'success',
                'info': item.type === 'info',
                'bare': item.type === 'bare',
                'rtl-mode': rtl
            }"

            (mouseenter)="onEnter()"
            (mouseleave)="onLeave()">

            <div *ngIf="!item.html">
                <div class="sn-title">{{item.title}}</div>
                <div class="sn-content">{{item.content | max:maxLength}}</div>

                <div class="icon" *ngIf="item.icon !== 'bare'" [innerHTML]="safeSvg"></div>
            </div>
            <div *ngIf="item.html" [innerHTML]="item.html"></div>

            <div class="sn-progress-loader" *ngIf="showProgressBar">
                <span [ngStyle]="{'width': progressWidth + '%'}"></span>
            </div>

        </div>
    `,
        styles: [`
        .simple-notification {
            width: 100%;
            padding: 10px 20px;
            box-sizing: border-box;
            position: relative;
            float: left;
            margin-bottom: 10px;
            color: #fff;
            cursor: pointer;
            transition: all 0.5s;
        }

        .simple-notification .sn-title {
            margin: 0;
            padding: 0 50px 0 0;
            line-height: 30px;
            font-size: 20px;
        }

        .simple-notification .sn-content {
            margin: 0;
            font-size: 16px;
            padding: 0 50px 0 0;
            line-height: 20px;
        }

        .simple-notification .icon {
            position: absolute;
            box-sizing: border-box;
            top: 0;
            right: 0;
            width: 70px;
            height: 70px;
            padding: 10px;
        }

        .simple-notification .icon svg {
            fill: #fff;
            width: 100%;
            height: 100%;
        }

        .simple-notification.rtl-mode {
            direction: rtl;
        }

        .simple-notification.rtl-mode .sn-content {
            padding: 0 0 0 50px;
        }

        .simple-notification.rtl-mode svg {
            left: 0;
            right: auto;
        }

        .simple-notification.error { background: #F44336; }
        .simple-notification.success { background: #8BC34A; }
        .simple-notification.alert { background: #ffdb5b; }
        .simple-notification.info { background: #03A9F4; }

        .simple-notification .sn-progress-loader {
            position: absolute;
            top: 0;
            left: 0;
            width: 100%;
            height: 5px;
        }

        .simple-notification .sn-progress-loader span {
            float: left;
            height: 100%;
        }

        .simple-notification.success .sn-progress-loader span { background: #689F38; }
        .simple-notification.error .sn-progress-loader span { background: #D32F2F; }
        .simple-notification.alert .sn-progress-loader span { background: #edc242; }
        .simple-notification.info .sn-progress-loader span { background: #0288D1; }
        .simple-notification.bare .sn-progress-loader span { background: #ccc; }
    `]
    }),
    __metadata("design:paramtypes", [notifications_service_1.NotificationsService,
        platform_browser_1.DomSanitizer,
        core_1.NgZone])
], NotificationComponent);
exports.NotificationComponent = NotificationComponent;
