// angular2-semantic-ui/components/accordion/accordion.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@angular/core");
const accordion_panel_1 = require("./accordion_panel");
let AccordionComponent = class AccordionComponent {
    constructor() {
        this.option = {};
        this._accordions = [];
    }
    getCls() {
        let cls = {
            "styled": this.option.styled == undefined ? true : this.option.styled,
            "fluid": this.option.fluid == undefined ? true : this.option.fluid,
            "inverted": this.option.inverted == undefined ? false : this.option.inverted
        };
        return cls;
    }
    setAccordion(accordion) {
        let isActive = accordion.active;
        if (!this.option.allowMultiple) {
            this._accordions.forEach(a => a.active = false);
        }
        accordion.active = !isActive;
    }
    ngAfterContentInit() {
        this._accordions = this.accordions.toArray();
        var self = this;
        this._accordions.forEach(a => {
            a.onChange(self);
        });
        if (!this.option.allowMultiple) {
            let finded = {};
            for (var i = 0; i < this._accordions.length; i++) {
                var item = this._accordions[i];
                if (item.active) {
                    item.active = false;
                    finded = item;
                }
            }
            if (finded.active !== undefined) {
                finded.active = true;
            }
        }
    }
};
__decorate([
    core_1.ContentChildren(accordion_panel_1.AccordionPanelComponent),
    __metadata("design:type", core_1.QueryList)
], AccordionComponent.prototype, "accordions", void 0);
__decorate([
    core_1.Input(),
    __metadata("design:type", Object)
], AccordionComponent.prototype, "option", void 0);
AccordionComponent = __decorate([
    core_1.Component({
        selector: 'lsu-accordion',
        template: `
    <div class="ui styled accordion" [ngClass]="getCls()">      
      <ng-content select="lsu-accordionPanel"></ng-content>
    </div>
  `
    }),
    __metadata("design:paramtypes", [])
], AccordionComponent);
exports.AccordionComponent = AccordionComponent;
