<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html><head>
<meta content="text/html; charset=ISO-8859-1" http-equiv="content-type"><title>
de novo
</title>

</head>
<body>
<h1 style="color: black;">
Detection of <em>de novo</em> variants
</h1>

<a href="#description">Description</a>&nbsp;&nbsp;
<a href="#prerequisites">Prerequisites</a>&nbsp;&nbsp;
<a href="#dialog">Dialog</a>&nbsp;&nbsp;
<a href="#tips">Practical tips</a>&nbsp;&nbsp;
<a href="#example">Example</a>&nbsp;&nbsp;
<a href="#algorithm">Algorithm</a>

<h2 id="description" style="color: black;">Description</h2><p>
Identification of <em>de novo</em> variants in a trio (child + both parents). The output includes posterior <em>de novo</em> probabilities 
as well as the percentage of reads with the alternative allele.
</p>

<h2 id="prerequisites" style="color: black;">Prerequisites</h2><p>
To apply the <em>de novo</em>
algorithm in FILTUS, the following requirements should be met:
<ul>
<li>The variants of the trio should be contained in <span style="font-style: italic;">a single file</span> containing all three samples. Joint variant
calling is highly recommended.</li>
<li>The variant file must have VCF-like genotype
columns, with the correct FORMAT column indicated in the <a href="loading.html#inputsettings">input settings dialog</a>.</li>
<li>The FORMAT column (the one with entries like
GT:AD:DP:GQ:PL) must contain the fields GT, AD and PL, which
are all used in the algorithm.</li>
<li>The variant file must be loaded with the <span style="font-style: italic;">keep 0/0</span> option
checked in the <a href="loading.html#inputsettings">input
settings dialog</a>.</li>
</ul>

<p><em>NOTE: If your variant files don't meet the above requirements, there may still be hope! In many cases you can identify potential <em>de novo</em> variants using a 
family based gene sharing with a dominant model. For instance, if all you have are individual variant files for a child and both parents, 
this would be the thing to do. The downside is that you wouldn't get posterior probabilities - or any other measure of classification strength - and probably quite a few more false positives.
</em></p>

<h2 id="dialog" style="color: black;">Dialog</h2>
<div style="text-align: center;"><img alt="De novo dialog" height="340" src="pictures/denovo.png"><br></div>

<p>To open the <em>de novo</em>
dialog, choose <span style="font-style: italic;">De novo
variant detection</span> in the <span style="font-style: italic;">Analysis</span> menu. Any
filters (e.g. PASS) should be applied before opening the dialog. The entry points of the dialog are as follows:
</p>

<dl>
<dt><strong>
Trio samples
</strong></dt><dd>
Indicate the sample numbers (corresponding to the sample
order in the <span style="font-style: italic;">Loaded
samples</span> window in the main FILTUS area.) Alternatively,
you can use the syntax <kbd>ID &lt;string&gt;</kbd> where
<kbd>&lt;string&gt;</kbd> is a unique identifier for the sample name (as
given in the variant file). For example, if the sample name of the
child is "Trio1_child" you can write <kbd>ID child</kbd>. If the string is not
unique (i.e. if there are multiple loaded sample names containing
"child") you will be warned. If you don't know who's who in the trio, 
you can find out using the <a href="qcplots.html#tips">Quality Control Plots</a> (see Tip 3 on that page).<br><br>
Finally, you should also indicate the gender of the child, to ensure correct handling of X-linked variants.
</dd><br>
<dt><strong>
Mutation rate
</strong></dt><dd>
This is used in the algorithm as the prior probability of a mutation at
a given position in a single meiosis. Default value: 1e-8. The
algorithm is usually not very sensitive to this parameter, but the
posterior probabilities will be affected if you change it radically.
</dd><br>

<dt><strong>
Allele frequencies
</strong></dt><dd>
Indicate a column containing frequencies for
the ALT alleles. The <span style="font-style: italic;">Missing
entry value</span> is substituted whenever the column does not
contain a number, or if the variant is multiallelic. If your variant file does not have frequency data,
the <span style="font-style: italic;">Missing entry value</span>
will be used for all variants. Without correct frequencies, the program
will still identify the same potential <em>de novo</em> variants,
but the posterior probabilities may be less accurate.</dd><br>

<dt><strong>
Output filters
</strong></dt><dd>
The purpose of these filters is to reduce the
number of false positives in the output. A true heterozygous <em>de novo</em>
variant is expected to present with ALT percentage close to 50% in the
child, and 0% in the parents. In practice some slack is recommended,
e.g. child &gt; 30% and parents &lt; 5%.<br><br>
One may experiment
with these filters for other purposes too: For example, to look for <em>de novo</em> 
mosaic variants in the child one could try a very loose cutoff for
the child, e.g. child &gt; 10%, while requiring parents = 0%. Or
oppositely, for variants inherited from a mosaic parent, something like
parents &lt; 25% and child &gt; 40% would be sensible, without
resulting in too many false positives. Of course, these are merely
suggestions whose validity depend heavily on the actual
contents of the variant file (e.g. the quality of the
variants and the parameters of the variant calling). </dd><br>

<dt><strong>
Summary
</strong></dt><dd>
A summary of the findings is printed here. The identified variants are shown in the main FILTUS window; to inspect them you must close the <em>de novo</em> dialog. 
</dd></dl>

<h2 id="tips" style="color: black;">Practical tips</h2>
<p><em>Tip 1:</em> 
To save the results, first close the <em>de novo</em> dialog, and then select Save main window content in the File menu.</p>
<p><em>Tip 2:</em> 
When browsing variants in the main FILTUS window, you can right click on any particular variant to see details about that variant for all the samples.</p>

<h2 id="example" style="color: black;">Example</h2>
<p>Among the testfiles included in the FILTUS package is the file <em>trioHG002_22X.vcf</em>, 
which contains variants on chromosomes 22 and X from the exome sequencing of a trio. The variants are annotated with Annovar.
(The trio is the Jewish trio from the Genome in a Bottle Consortium, described <a href="http://biorxiv.org/content/early/2015/09/15/026468">here</a>.
The complete exome sequencing variant set, is publically available from this
<a href="ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio/analysis/OsloUniversityHospital_Exome_GATK_jointVC_11242015">ftp site</a>.)<br>
<br>
The following steps shows a typical <em>de novo</em> analysis in FILTUS.
<ol>
<li>Load the <em>trioHG002_22X.vcf</em> file in the <em>testfiles</em> folder. Make sure the "Keep 0/0" box is checked.
</li>
<li>To remove the worst noice, apply column filters "VCF_FILTER - equal to - PASS" and "DP - greater than - 9".
</li>
<li>Find out who's who in the trio: Open the QC plot dialog (in the <em>Analysis</em> menu) and just press the first "Plot!" button.
The Private variants plot should convince you that sample 1 is the child, which in combination with the Gender plot implies that the sample order is
1=boy, 2=father, 3=mother.
</li>
<li>Close the plot dialog and open the <em>de novo</em> dialog. Enter the sample order and child gender as deduced above, 
and choose "1000g2014oct_all" as the allele frequency column.</li>
<li>Press Compute.
</li></ol>
If you didn't set any output thresholds, the summary display should say that 4 variants were found. To examine them, first close the dialog. 
By right clicking any individual variant you'll see genotype details from all loaded samples about that variant.<br><br>
A few words about interpreting the findings: 
<ul>
<li>The top two variants are listed with posterior probability 1. However, they both have lower ALT percentage (ca 38%) 
than you would expect from a heterozygous <em>de novo</em> variant. The fact they are only 52 basepairs apart is a bit suspicious, 
and we see from the <em>genomicSuperDups</em> column that they are in a region of segmental duplication. 
All in all, it is quite likely that both of these are false positives.</li>
<li>The third variant has posterior probability 0, and is almost certainly a false positive. This is a tri-allelic variant, 
where the TAA > TA change is reported to be <em>de novo</em>. If you right click on the variant you will see that both parents actually has
plenty of reads with the alledged <em>de novo</em> allele.
(Note about the alleles: Don't look at the "Ref" and "Alt" columns, which are added by Annovar and do not preserve multiple alternative alleles in this file. 
The correct allele columns are "VCF_REF" and "VCF_ALT" much further to the right.)
</li>
<li>The ALT percentages of the fourth variant look very convincing, so why does it have posterior probability 0? 
The problem is the father: According to the details, he has 365 REF reads and 0 ALT, but for some reason the genotype quality GQ is 0. 
This could be a mapping problem (perhaps related to the fact that the variant is a fairly long insertion - or segmental duplication), or something else. 
Anyway it would be worth investigating closer.</li>
</ul>
In summary: 1) Posterior probabilities should be taken with a grain of salt, and 
2) If you have access to the BAM files, a closer look at the actual reads (e.g. in IGV or similar software) is always a good idea.


<h2 id="algorithm" style="color: black;">Algorithm</h2><p>
FILTUS uses the GT field of the genotype columns to recognize the following <em>de novo</em> genotype patterns (father + mother = child):
<ul>
<li>Autosomal:
    <ul>
    <li>0/0 + 0/0 = 0/1</li>
    <li>0/0 + 0/0 = 1/1</li>
    <li>0/0 + 0/1 = 1/1</li>
    <li>0/1 + 0/0 = 1/1</li>
    </ul></li>
<li>X-linked, child is boy:
    <ul>
    <li>0 + 0/0 = 1</li>
    </ul></li>
<li>X-linked, child is girl:
    <ul>
    <li>0 + 0/0 = 0/1</li>
    <li>0 + 0/0 = 1/1</li>
    <li>0 + 0/1 = 1/1</li>
    </ul></li>
</ul>
A variant is treated as X-linked in this context only if it is located outside of the pseudoautosomal regions PAR1 and PAR2 on the X chromosome. 
Multiallelic generalizations of the above patterns are also caught. 
However, combinations with any of the following properties are treated as benign and discarded from further analyses:
<ul>
<li>The <em>de novo</em> allele is 0 (= REF). Example: 1/1 + 1/1 = 0/1.</li>
<li>Child genotype equals either of the parents. Example: 0/0 + 1/1 = 1/1.</li>
<li>Missing genotype in any trio member.</li>
<li>A male trio member is reported as heterozygous for an X-linked variant.</li>
</ul>
For each variant where a <em>de novo</em> genotype combination is identified, FILTUS reports the posterior <em>de novo</em> probability. This is computed using a Bayesian formula involving the 
population allele frequencies, the genotype likelihoods reported by the variant caller (in the PL fields) and the prior mutation rate specified by the user.
</p>
</body></html>