<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html><head>
<meta content="text/html; charset=ISO-8859-1" http-equiv="content-type"><title>filtus</title></head>
<body>
<h1 style="color: black;">FILTUS</h1>
<big>FILTUS is a stand-alone tool for downstream
analysis in high-throughput sequencing projects. It is especially well
suited for identification of variants causing Mendelian disease. The program is written in Python, by:
<br><br>Magnus Dehli Vigeland, PhD<br>
Department of Medical Genetics<br>
Oslo University Hospital, Norway<br>
Email: magnusdv&lt;at&gt;medisin&lt;dot&gt;uio&lt;dot&gt;no</big><br><big>
<br>FILTUS accepts (at least in principle) any format of variant files, as
long as you state which columns contain the essential
information (e.g. chromosome, position, gene and genotype). Standard
VCF files are automatically recognized. Unlike other variant tools, FILTUS is completely self-contained and
does not require an internet connection. This may be an important point
if you are not allowed to be online when working with human
sequence data.<br>
</big><span style="font-weight: bold;"><br>
</span><big>The latest version of FILTUS is 1.0.5. Take a
look at the <a href="FILTUS/Filtus_changelog.txt" target="_top">changelog</a>
to see recent developments.</big><big> Questions,
bug reports, feature requests and other comments are welcome by email
to magnusdv&lt;at&gt;medisin&lt;dot&gt;uio&lt;dot&gt;no.</big><br>

<hr style="width: 100%; height: 2px;"><h2><span style="font-weight: bold;">Features</span></h2>
<ul>
<li>Overview statistics of files and of individual columns.</li>
<li>Filters: </li>
<ul>
<li>Any number of column filters</li>
<li>Exclude variants in inhouse datase</li>
<li>Exclude/restrict to genes in a gene list</li>
<li>Restrict to certain genomic regions</li>
</ul>
<li>Breakdown table of the filtering process.</li><li>De novo analysis in trios.</li><li>Autozygosity mapping in individuals with known or suspected inbreeding ("AutEx").</li><li>Gene based analysis: Identify genes where specified samples
contain variants.</li>
<li>Genetic models: Dominant, simple recessive, compound
heterozygous.</li>
<li>Statistical model for prioritizing genes in Mendelian
disease mapping projects.</li>
<li>Sporadic patients or family based analysis.</li>

<li>Export to MERLIN-format (for linkage analysis).</li>
<li>Simultaneous analysis of up to several hundred complete
exomes.</li><li>Prefilter files when loading: Allows analysis of whole-genome size files.</li>
<li>Quality control plots:</li>
<ul>
<li>Compare heterozygosity levels, private variants and other
features.</li>
<li>Scatter plots of any numerical columns
- stratified on genotype.</li>
</ul>
<li>Reproducibility: Save/load filters. </li>

</ul>

<hr style="width: 100%; height: 2px;">

<h2>Download and installation</h2>
<strong>WINDOWS</strong><br>
<ul>
<li>Download the latest version <a href="FILTUS/Filtus1.0.5.zip">here</a>
and unpack the zip-file in any location you prefer. To start the
program, simply double-click on the file FiltusXXX.exe in the
unpacked folder (XXX = version number).</li>
</ul>
A couple of things to notice:<br>
<ol>
<li>The windows version of FILTUS requires a 64-bit processor.</li>
<li>The unpacked Filtus folder contains a great number of
files in addition to the .exe file. These are for internal use only and
should be ignored (but not deleted) by end users.</li>
<li>To create a FILTUS icon on the desktop (Windows 7):</li>
<ol>
<li>Right click on the .exe file</li>
<li>Choose <em>Send to</em> and then <em>Desktop
(create shortcut)</em></li>
</ol>
<li>To uninstall FILTUS you can just delete the Filtus folder.
FILTUS leaves no footprint on your computer.</li>
</ol>
<br>
<h3>MAC/LINUX</h3>
<ul>
<li>If you have pip installed, automatic installation can be done by <span style="font-family: Courier New;">pip install filtus</span> in a terminal.</li>
<li>Alternatively, you are welcome to download the source files
and run Filtus from a command line:
<ul>
<li>Download the source files <a href="FILTUS/Filtus1.0.5_SOURCE.zip">here</a> and unpack
the zip-file in a location of your choice.</li>
<li>Open a terminal and move to the unpacked Filtus folder.</li>
<li>To start the program, use the command <span style="font-family: Courier New;">python Filtus.py</span><br>
</li>
</ul>
</ul>
Requirements when running FILTUS from source files:<br>
<ol>
<li>Python version 2.7.x where x is at least 3. (Get
it from <a href="http://www.python.org/download/">python.org</a>
if you don't have it.) In particular, Python 3 will not work.  </li>
<li>For the plotting functionality (QC plots) to work, the two
Python packages <a href="http://sourceforge.net/projects/numpy/">numpy</a>
and <a href="http://matplotlib.org/downloads.html">matplotlib</a>
must be installed on your system.</li>
</ol>
<br>
<hr style="width: 100%; height: 2px;">
<h2>Loading a test file</h2>
To check that everything is OK, you can load a test file as follows:<br>
<ul>
<li>Start FILTUS.</li>
<li>On the <em>File</em> menu,
choose <em>Load files (simple).</em></li>
<li>Move to the <em>testfiles</em>
folder, choose <em>test_file.csv</em> and
press <em>Open.</em></li>
<li>In the Input file settings dialog, click <em>Use
for all files.</em></li>
<li>You should now see the file name in the <em>Loaded
samples</em> window, and a row in the <em>Unfiltered
summaries</em> window saying "425 variants in 205 genes". If you
double-click on this row, you should see the contents of the file in
the main window.</li>
</ul>
</body></html>